### PR TITLE
CMake: Revise status output for important build options & fix misc. annoyances

### PR DESCRIPTION
**Pull request by kinke**

### CODE REVIEW

**Comment by Geod24** [comment]
I gave this a try on Mac OSX, the LLVM detection works fine but the first run complained about `ldmd2`:
```console
geod24@geod24mbp2018 ~/projects/dlang/ldc/build-test (git)-[remotes/kinke/cmake_misc] % cmake ..
-- The C compiler identification is AppleClang 12.0.0.12000032
-- The CXX compiler identification is AppleClang 12.0.0.12000032
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found host D compiler /usr/local/bin/ldmd2, with default flags ''
-- Host D compiler ID: LDMD
-- Host D compiler version: LDC - the LLVM D compiler (1.24.0)
-- Host D compiler front-end version: 2094
-- LLVM_NATIVE_ARCH: X86
-- Found LLVM: /usr/local/Cellar/llvm/11.0.1 (found suitable version "11.0.1", minimum required is "6.0")
-- Found mlir-tblgen: /usr/local/Cellar/llvm/11.0.1/bin/mlir-tblgen
-- -- Building LDC with MLIR support (/usr/local/Cellar/llvm/11.0.1)
-- Found PkgConfig: /usr/local/bin/pkg-config (found version "0.29.2")
-- Checking for module 'LLVMSPIRVLib'
--   No package 'LLVMSPIRVLib' found
-- -- Building LDC without SPIR-V support: not found
-- Performing Test SUPPORTS_FVISIBILITY_INLINES_HIDDEN_FLAG
-- Performing Test SUPPORTS_FVISIBILITY_INLINES_HIDDEN_FLAG - Success
-- LDC version identifier: 1.25.0-git-a7f9cf7-dirty
-- -- Building LDC with dynamic compilation support (LDC_DYNAMIC_COMPILE): ON
-- -- Building LDC with integrated LLD linker (LDC_WITH_LLD): ON
-- -- Building LDC with plugin support (LDC_ENABLE_PLUGINS): ON
CMake Error at cmake/Modules/ExtractDMDSystemLinker.cmake:41 (message):
  Failed to link empty D program using '/usr/local/bin/ldmd2 -wi'
Call Stack (most recent call first):
  CMakeLists.txt:642 (include)


-- Configuring incomplete, errors occurred!
See also "/Users/geod24/projects/dlang/ldc/build-test/CMakeFiles/CMakeOutput.log".
See also "/Users/geod24/projects/dlang/ldc/build-test/CMakeFiles/CMakeError.log".
```

When I re-ran it, without changing anything, it passed:
```console
1 geod24@geod24mbp2018 ~/projects/dlang/ldc/build-test (git)-[remotes/kinke/cmake_misc] % cmake ..                                                                                                                                         :(
-- Found host D compiler /usr/local/bin/ldmd2, with default flags ''
-- Host D compiler ID: LDMD
-- Host D compiler version: LDC - the LLVM D compiler (1.24.0)
-- Host D compiler front-end version: 2094
-- LLVM_NATIVE_ARCH: X86
-- Found mlir-tblgen: /usr/local/Cellar/llvm/11.0.1/bin/mlir-tblgen
-- -- Building LDC with MLIR support (/usr/local/Cellar/llvm/11.0.1)
-- Checking for module 'LLVMSPIRVLib'
--   No package 'LLVMSPIRVLib' found
-- -- Building LDC without SPIR-V support: not found
-- LDC version identifier: 1.25.0-git-a7f9cf7-dirty
-- -- Building LDC with dynamic compilation support (LDC_DYNAMIC_COMPILE): ON
-- -- Building LDC with integrated LLD linker (LDC_WITH_LLD): ON
-- -- Building LDC with plugin support (LDC_ENABLE_PLUGINS): ON
-- Host D compiler linker program: /usr/bin/cc
-- Host D compiler linker flags: -L/usr/local/Cellar/ldc/1.24.0/lib;-lphobos2-ldc;-ldruntime-ldc;-ldl;-lpthread;-lm;-m64
-- -- Including LTO linker plugin (LDC_INSTALL_LTOPLUGIN): ON (/usr/local/Cellar/llvm/11.0.1/lib/libLTO.dylib)
-- -- Including LLVM compiler-rt libraries (LDC_INSTALL_LLVM_RUNTIME_LIBS): ON
-- --  - /usr/local/Cellar/llvm/11.0.1/lib/clang/11.0.1/lib/darwin/libclang_rt.asan_osx_dynamic.dylib --> libldc_rt.asan.dylib
/Library/Developer/CommandLineTools/usr/bin/install_name_tool: warning: changes being made to the file will invalidate the code signature in: /Users/geod24/projects/dlang/ldc/build-test/lib/libldc_rt.asan.dylib (for architecture x86_64)
/Library/Developer/CommandLineTools/usr/bin/install_name_tool: warning: changes being made to the file will invalidate the code signature in: /Users/geod24/projects/dlang/ldc/build-test/lib/libldc_rt.asan.dylib (for architecture arm64)
-- --  - /usr/local/Cellar/llvm/11.0.1/lib/clang/11.0.1/lib/darwin/libclang_rt.tsan_osx_dynamic.dylib --> libldc_rt.tsan.dylib
/Library/Developer/CommandLineTools/usr/bin/install_name_tool: warning: changes being made to the file will invalidate the code signature in: /Users/geod24/projects/dlang/ldc/build-test/lib/libldc_rt.tsan.dylib (for architecture x86_64)
/Library/Developer/CommandLineTools/usr/bin/install_name_tool: warning: changes being made to the file will invalidate the code signature in: /Users/geod24/projects/dlang/ldc/build-test/lib/libldc_rt.tsan.dylib (for architecture arm64)
-- --  - /usr/local/Cellar/llvm/11.0.1/lib/clang/11.0.1/lib/darwin/libclang_rt.osx.a --> libldc_rt.builtins.a
-- --  - /usr/local/Cellar/llvm/11.0.1/lib/clang/11.0.1/lib/darwin/libclang_rt.profile_osx.a --> libldc_rt.profile.a
-- --  - /usr/local/Cellar/llvm/11.0.1/lib/clang/11.0.1/lib/darwin/libclang_rt.fuzzer_osx.a --> libldc_rt.fuzzer.a
-- --  - /usr/local/Cellar/llvm/11.0.1/lib/clang/11.0.1/lib/darwin/libclang_rt.xray_osx.a --> libldc_rt.xray.a
-- --  - /usr/local/Cellar/llvm/11.0.1/lib/clang/11.0.1/lib/darwin/libclang_rt.xray-basic_osx.a --> libldc_rt.xray-basic.a
-- --  - /usr/local/Cellar/llvm/11.0.1/lib/clang/11.0.1/lib/darwin/libclang_rt.xray-fdr_osx.a --> libldc_rt.xray-fdr.a
-- --  - /usr/local/Cellar/llvm/11.0.1/lib/clang/11.0.1/lib/darwin/libclang_rt.xray-profiling_osx.a --> libldc_rt.xray-profiling.a
-- Using path for Intrinsics.td: /usr/local/Cellar/llvm/11.0.1/include
-- The ASM compiler identification is Clang
-- Found assembler: /Library/Developer/CommandLineTools/usr/bin/cc
-- -- LDC runtime configuration:
-- --  - Building 32/64-bit libraries (MULTILIB): OFF
-- --  - Building shared libraries (BUILD_SHARED_LIBS): BOTH
-- --  - Building LTO libraries (BUILD_LTO_LIBS): OFF
-- Looking for unistd.h
-- Looking for unistd.h - found
-- Use custom passes in jit: ON
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/geod24/projects/dlang/ldc/build-test
```

**Comment by kinke** [comment]
Thx for testing; please re-test again from scratch, as that error should now show the compiler stderr output for troubleshooting.

**Comment by Geod24** [comment]
```console
-- LDC version identifier: 1.25.0-git-fbc3161
-- -- Building LDC with dynamic compilation support (LDC_DYNAMIC_COMPILE): ON
-- -- Building LDC with integrated LLD linker (LDC_WITH_LLD): ON
-- -- Building LDC with plugin support (LDC_ENABLE_PLUGINS): ON
CMake Error at cmake/Modules/ExtractDMDSystemLinker.cmake:41 (message):
  Failed to link empty D program using '/usr/local/bin/ldmd2 -wi':

  ld: library not found for -ldl

  clang: error: linker command failed with exit code 1 (use -v to see
  invocation)

  Error: /Library/Developer/CommandLineTools/usr/bin/cc failed with status: 1
Call Stack (most recent call first):
  CMakeLists.txt:642 (include)


-- Configuring incomplete, errors occurred!
See also "/Users/geod24/projects/dlang/ldc/build-test/CMakeFiles/CMakeOutput.log".
See also "/Users/geod24/projects/dlang/ldc/build-test/CMakeFiles/CMakeError.log".
```

**Comment by kinke** [comment]
Hmm, that would mean a trivial `/usr/local/bin/ldmd2 -wi -v hello.d` fails to link, with the src file containing `void main() {}`. Can you investigate a bit? Possibly outputting `${stdout}` too so that it should become clear which config file it uses and how it invokes clang.

**Comment by Geod24** [comment]
Hum, the issue is that CMake is trying to pull in `libdl` which doesn't exist on Mac (see e.g. https://reviews.llvm.org/D43168).
On the first run, cmake detects the Developer tools (aka clang) and clang just errors out. When re-running cmake, it picks up `/usr/bin/cc`, which is a compatibility layer (acts like GCC) and either just ignores `-ldl` or links an empty lib.

However, I don't know why `-ldl` is used. `DFLAGS`, `LDFLAGS`, `LINK_FLAGS` are empty.

**Comment by kinke** [comment]
Hmm, the compiler explicitly adds it for Darwin: https://github.com/ldc-developers/ldc/blob/4bee4e6f2033d6a84d3cbdd1aec4d464a0bb5c58/driver/linker-gcc.cpp#L650-L658

File: runtime/DRuntimeIntegrationTests.cmake [hunk 1]
```diff
@@ -40,9 +40,8 @@ else()
     set(cflags_base "CFLAGS_BASE=-Wall -Wl,-rpath,${CMAKE_BINARY_DIR}/lib${LIB_SUFFIX}")
 endif()
 
-if("${TARGET_SYSTEM}" MATCHES "FreeBSD|DragonFly")
-    set(linkdl "")
-else()
+set(linkdl "")
+if("${TARGET_SYSTEM}" MATCHES "Linux")
     set(linkdl "LINKDL=-L-ldl")
 endif()
```
Comment (Geod24) [inline runtime/DRuntimeIntegrationTests.cmake:46]: Wouldn't https://cmake.org/cmake/help/latest/variable/CMAKE_DL_LIBS.html be better here ?

Comment (kinke) [reply runtime/DRuntimeIntegrationTests.cmake:46]: Not for cross-compilation with our own `TARGET_SYSTEM` semantics.

**Comment by kinke** [comment]
@Geod24: Have you tested the latest version by chance?

**Comment by Geod24** [comment]
@kinke : Sorry missed the notification. Haven't tested recently (since it needed to be compiled with a different host).
Will update the Alpine / Homebrew package and let you know if I find any issue.

**Comment by kinke** [comment]
No worries; thx for updating and note that you might be affected by #3673 as package maintainer (if so, then hopefully positively :)).

**Pull request merged**